### DISCUSSION TITLE
Doc on how to write tests

### DISCUSSION THREAD

**Discussion by SylvainMartel** [tokio-rs/axum#555] [Q&A] [answered]
Is there any doc on how to write tests with axum?  I saw the example but it's not enough for a novice like me :)  I was able to write simple route tests, but as soon as I try to write tests that access the database with sqlx, I am faced with "Future"  errors that I can't solve.  I looked at the API doc and here in the repo for some guidance/documentation, but beside this [example](https://github.com/tokio-rs/axum/tree/main/examples/testing) I cannot find anything.  

Any tips on where to look?
Is testing done with something external to axum?  

Thanks!

**Comment by davidpdrsn** [comment]
We don't have anything outside of that example as that requires documenting testing each and every individual library.

What are the specific "future" issues you have?

**Comment by SylvainMartel** [comment, marked as answer]
Here is the modified app fn after I added the pgpool.  I added "async" so that the pool would work.  But this will give me an error for the line `.serve(app().into_make_service())` 

> no method named `into_make_service` found for opaque type `impl Future` in the current scope
> method not found in `impl Future`



```
#[allow(dead_code)]
async fn app() -> Router {
    dotenv().ok();
    let database_url = env::var("DATABASE_URL").expect("DATABASE_URL is not set in .env file");
    let db_pool = PgPool::connect(&database_url).await.unwrap();
    let shared_state = Arc::new(AppState {
        health_check_response: "I'm good.  You've already asked me ".to_string(),
        visit_count: Mutex::new(0),
        db: db_pool,
    });
    Router::new()
        .route("/", get(root))
        .route("/health", get(health_check_handler))
        .route("/courses", post(post_new_course))
        .route("/courses/:tutor_id", get(get_courses_for_tutor))
        .route("/courses/:tutor_id/:course_id", get(get_course_details).put(update_course_details).delete(delete_course))
        .route("/tutors", post(post_new_tutor).get(get_all_tutors))
        .route("/tutors/:tutor_id", delete(delete_tutor).get(get_tutor_details)) //put(update_tutor_details).
        .layer(AddExtensionLayer::new(shared_state))
        // We can still add middleware
        .layer(TraceLayer::new_for_http())
}

#[cfg(test)]
mod tests {
    use super::*;
    use axum::{
        body::Body,
        http::{self, Request, StatusCode},
    };
    use serde_json::{json, Value};
    use std::net::{SocketAddr, TcpListener};
    use tower::ServiceExt; // for `app.oneshot()`


    
    
    #[tokio::test]
    async fn the_real_deal() {
        let listener = TcpListener::bind("127.0.0.1:8000".parse::<SocketAddr>().unwrap()).unwrap();
        let addr = listener.local_addr().unwrap();

        tokio::spawn(async move {
            axum::Server::from_tcp(listener)
                .unwrap()
                .serve(app().into_make_service())
                .await
                .unwrap();
        });

        let client = hyper::Client::new();

        let response = client
            .request(
                Request::builder()
                    .uri(format!("http://{}", addr))
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();

        let body = hyper::body::to_bytes(response.into_body()).await.unwrap();
        assert_eq!(&body[..], b"Hello, World!");
    }
}
```

**Reply by davidpdrsn** [reply]
Seems to me you're missing a `.await`:

```diff
-.serve(app().into_make_service())
+.serve(app().await.into_make_service())
```

**Reply by SylvainMartel** [reply]
*bonk head*
Thanks, I'm still learning to understand the compiler error the hard way it seems.  
This also solved using the functions without starting a server and I was able to make a successful test that query the database. 

Thanks for taking the time! :)

**Reply by davidpdrsn** [reply]
Np :)

**Comment by renancloudwalk** [comment]
@kinwolfqc I'm aware that this issue is stale, but I created a lib based on TestClient (that is inside axum), it reduces most of the boilerplate for testing axum:
https://crates.io/crates/axum-test-helper

remmember to add it inside dev-dependencies, instead of dependencies directly
